File: plugin/src/core/policy.ts
export interface Decision {
  accepted: Set<string>;
  held: Set<string>;
  ignored: Set<string>;
}

export interface DecideOptions {
  autoThreshold: number;
  triageThreshold: number;
  suppressed?: Set<string>;
}

export function decide(
  scores: Record<string, number>,
  options: DecideOptions
): Decision {
  const { autoThreshold, triageThreshold, suppressed = new Set<string>() } = options;
  const accepted = new Set<string>();
  const held = new Set<string>();
  const ignored = new Set<string>();

  for (const [tag, confidence] of Object.entries(scores)) {
    if (suppressed.has(tag) || confidence < triageThreshold) {
      ignored.add(tag);
    } else if (confidence >= autoThreshold) {
      accepted.add(tag);
    } else {
      held.add(tag);
    }
  }

  return { accepted, held, ignored };
}

File: src/zotero_organiser/policy.py
from __future__ import annotations

from collections.abc import Set
from dataclasses import dataclass


@dataclass(frozen=True)
class Decision:
    accepted: set[str]
    held: set[str]
    ignored: set[str]


def decide(
    scores: dict[str, float],
    *,
    auto_threshold: float,
    triage_threshold: float,
    suppressed: Set[str] = frozenset(),
) -> Decision:
    accepted, held, ignored = set(), set(), set()
    for tag, confidence in scores.items():
        if tag in suppressed or confidence < triage_threshold:
            ignored.add(tag)
        elif confidence >= auto_threshold:
            accepted.add(tag)
        else:
            held.add(tag)
    return Decision(accepted, held, ignored)
